### PR TITLE
manubot-pandoc-cite manubot-bibliography-cache

### PR DESCRIPTION
**Pull request by dhimmel**
refs https://github.com/manubot/manubot/issues/258

### CODE REVIEW

**Comment by dhimmel** [comment]
@slippycheeze this PR enables the metadata variable `manubot-bibliography-cache`. It defaults to None / null / no caching, since I'm not sure there is a default path that works well for all users. We can always revisit the default decision in the future and either add one or emit a warning if no cache is specified.

Also added support for YAML output (detected by `.yaml` or `.yml` path suffix). Still a bit experimental, but might be nice for human readability/editability.

File: manubot/cite/citations.py [hunk 1]
```diff
@@ -177,17 +178,36 @@ def csl_json(self) -> str:
...
+            default_flow_style=False,
+            width=float("inf"),
+            allow_unicode=True,
+            sort_keys=False,
+        )
+
+    def write_csl_items(self, path: tp.Union[os.PathLike, str, None]) -> None:
+        """
+        Write CSL Items to a JSON or YAML file at `path`.
+        If path ends with a .yml or .yaml extension, write as CSL YAML.
```
Comment (agitter) [inline manubot/cite/citations.py:200]: I like this strategy for YAML support.  Continue using JSON by default and support YAML optionally based on extension.

Comment (agitter) [inline manubot/cite/citations.py:200]: I like this strategy for YAML support. Continue using JSON by default and support YAML optionally based on extension.

File: manubot/pandoc/cite_filter.py [hunk 2]
```diff
@@ -154,19 +162,32 @@ def process_citations(doc: pf.Doc) -> None:
 
     - bibliography (use to define reference metadata manually)
     - citekey-aliases (use to define tags for cite-by-id citations)
+    - manubot-bibliography-cache:
```
Comment (agitter) [inline manubot/pandoc/cite_filter.py:165]: To make sure I'm following the usage correctly, to enable the desired caching would a user provide the same path to `manubot-bibliography-cache` and `manubot-output-bibliography`?  Or they could use different paths, but populate `manubot-bibliography-cache` by taking `manubot-output-bibliography` as a starting point and manually deleting any entries that should not be cached?

File: manubot/pandoc/cite_filter.py [hunk 3]
```diff
@@ -195,7 +216,8 @@ def process_citations(doc: pf.Doc) -> None:
         req_cache.close()
 
     citations.write_citekeys_tsv(path=doc.get_metadata("manubot-output-citekeys"))
-    citations.write_csl_json(path=doc.get_metadata("manubot-output-bibliography"))
+    citations.write_csl_items(path=doc.get_metadata("manubot-output-bibliography"))
+    citations.write_csl_items(path=doc.manubot["bibliography_cache"])
```
Comment (agitter) [inline manubot/pandoc/cite_filter.py:220]: I may not be following the behavior here.  All of the citations are written to both the output bibliography and the bibliography cache, overwriting the existing bibliography cache?

File: manubot/pandoc/tests/test_cite_filter/input.md [hunk 4]
```diff
@@ -11,6 +11,7 @@ references:
   container-title: Greene Laboratory
   location: University of Pennsylvania
   issued: {'date-parts': [[2018, 12, 19]]}
+# manubot-bibliography-cache: manubot/pandoc/tests/test_cite_filter/bibliography-cache.json
```
Comment (agitter) [inline manubot/pandoc/tests/test_cite_filter/input.md:14]: Why is this commented out?

File: manubot/util.py [hunk 5]
```diff
@@ -142,3 +143,35 @@ def read_serialized_dict(path: str) -> dict:
         f"Expected data encoded by {path!r} to be a dictionary at the top-level. "
         f"Received {data.__class__.__name__!r} instead."
     )
+
+
+def _yaml_str_representer(dumper: "yaml.Dumper", data: str):
+    """
+    Use YAML literal block style for multiline strings.
```
Comment (agitter) [inline manubot/util.py:150]: How universal is this literal block style?

**Review by agitter — COMMENTED** [review]
This is looking good to me.  My comments pertain to usage to make sure I understand how a user will interact with the cache and how that addresses the proposed behavior in #258.

How is the cache populated the first time?  Is the idea that the file will not exist initially so it will not impact the first run of the cite filter.  But then it will be written and available after that first run?

I'm leaning toward leaving the new bibliography cache off by default.  I haven't thought through how it will help or hurt documents with references that intermittently fail.  Without the cache, the failures may eventually resolve themselves once the remote data source is responsive.  However, with a cache could we end up caching and reusing the minimal metadata from a failed request (e.g. the URL only)?

Are there other design questions from #258 I should comment on here?

File: manubot/util.py (hunk 5 above)
Comment (dhimmel) [reply manubot/util.py:150]: I think any parser that fully implements the YAML 1.1 spec should be able to parse this.

At https://yaml-multiline.info/, the format we specify for multiline strings, matches the options under block scalars:

- Block Scalar Style: Keep newlines (literal)
- Block Chomping: No newline at end (strip)

From what I can tell, this format is the best if we envision the YAML being tracked with git.

File: manubot/pandoc/tests/test_cite_filter/input.md (hunk 4 above)
Comment (dhimmel) [reply manubot/pandoc/tests/test_cite_filter/input.md:14]: For this test, don't want to use a CSL-bibliography cache. Uses --bibliography instead.

File: manubot/pandoc/cite_filter.py (hunk 3 above)
Comment (dhimmel) [reply manubot/pandoc/cite_filter.py:220]: `manubot-output-bibliography` and `manubot-bibliography-cache` and both default to None. If both are specified, two bibliography files can be written. I don't think most users will specify both.

Will improve docs.

**Comment by dhimmel** [comment]
> I haven't thought through how it will help or hurt documents with references that intermittently fail. Without the cache, the failures may eventually resolve themselves once the remote data source is responsive. However, with a cache could we end up caching and reusing the minimal metadata from a failed request (e.g. the URL only)?

Great point. The fallback URL reference methods aren't really something we want to be caching. Perhaps in the future, we could set a field on the CSL Item custom field like cache=False.

**Pull request merged**